Feat : Profile, CommentInput, Comment Component 구현

Jungjjeong opened the pull request:
## 📌 설명
PostDetail Page에서 사용되는 댓글 창 구현을 위해 3개의 컴포넌트를 구현했습니다.
- Profile Component
![image](https://user-images.githubusercontent.com/72294509/182182859-fd30b434-4b2d-4071-a236-e0a20a355c88.png)

- CommentInput Component
![image](https://user-images.githubusercontent.com/72294509/182182963-2807a7cc-996a-4c43-baac-5b7029f2982c.png)

-  Comment Component
![image](https://user-images.githubusercontent.com/72294509/182183104-8b25a724-ca24-40aa-8311-3440fe7040cf.png)

<br /><br />

```
이용방법(/information) 페이지에서 확인 가능합니다!
```

## 🎨 구현 내용
### 1. Profile Component Props
```
interface Props {
  iconSize: number | string;
  version: 'comment' | 'profile';
  user: UserInfo;
  onClick?: MouseEventHandler;
  createdAt?: string;
}

const defaultProps = {
  iconSize: 100,
  version: 'profile',
};
```
- `version`을 선택해서 `profile, comment`로 스타일 변경이 가능합니다. 
다만 `iconSize`는 `default: 100px`, `comment`를 위해서 사이즈를 줄이기 위해서는 직접 변경해줘야 합니다. 
- onClick 이벤트는 아이콘과 닉네임에 적용됩니다. 

<br />

### 2. CommentInput Component Props
- Props는 없습니다. 
- `CommentInput Component`는 **오직 댓글을 작성하고, 등록하는 기능에만 사용**합니다. 
따라서, `handleClick` 함수를 props로 받지 않고, 내부 로직은 `CommentInput.tsx` 안에 직접 작성할 예정입니다.

<br />

### 3.  Comment Component Props
```
interface Props {
  comment: Comment;
}
```
- Comment 객체 내부에는 `id, content, user, createdAt, children` 등이 있습니다. (DummyData Type 파일 참고)

<br /><br />

## ✅ 중점적으로 봐줬으면 하는 부분
- 더미데이터를 일부 수정했습니다. 확인 부탁드립니다. (User 객체 introduce, Comment 객체 children)
- 내부의 답글이 들어있는 `children`에서도 id, content 등의 **동일한 Comment 객체 형태**를 따릅니다. 
따라서 **위에서 선언된 변수를 또 선언**한다고 eslint 경고 문구가 출력됩니다. 
![image](https://user-images.githubusercontent.com/72294509/182186417-319e5a08-f1b9-46a7-a16e-265b2074f3ee.png)

경고를 없애기 위해 구조분해할당을 이용하지 않고, `child.id` 와 같은 방법으로 작성하면 코드가 지저분해질 것 같습니다. 
다른 분들 의견이 어떠신지 궁금합니다. 

<br /><br />

## 🚀 연관된 이슈
Closes #36

Jungjjeong committed:
Feat: Profile Component 구현

Jungjjeong committed:
Feat: Avatar Component default icon 수정

Jungjjeong committed:
Fix: DummyData에 맞춰 Profile Component 수정

- DummyData UserInfo -> introduce 추가

Jungjjeong committed:
Feat: Comment Component 기초 구현

Jungjjeong committed:
DesignL MyFoldersAreaLogOut Component style 수정

Jungjjeong committed:
Feat: comment DummyData 수정

Jungjjeong committed:
Feat: Comment Component 답글 기능 추가

- 수정된 API에 맞게 답글 기능 추가

Jungjjeong committed:
Feat: CommentInput Component 구현

Jungjjeong committed:
Fix: Comment Component 버튼 위치 변경

Jungjjeong committed:
Fix: Comment Component key 추가

Jungjjeong committed:
Feat: Test를 위해 information 페이지에 Component들 추가

Jungjjeong committed:
Feat: Profile Component onClick Prop 추가

Jungjjeong committed:
Chore: 필요하지 않은 주석 삭제

Steadystudy commented on web/components/Comment/Comment.tsx:
```diff
@@ -0,0 +1,69 @@
+      <S.RepliesContainer>
+        {showInputArea && <CommentInput />}
+        {showReplies &&
+          children?.map((child) => {
```
child에 any로 타입받고있는데 `child: Comment`하면 될 거 같습니다

Jungjjeong replied:
으아닛 급하게 추가했습니다

Steadystudy reviewed (approved):
깔끔하게 잘 구현하신 거 같습니다.
type 부분에 오류가 뜨긴 하지만 더미데이터 타입이 임시다 보니 크게 중요하진 않을 거 같아요.
추후 api로 받아오는 res type을 확실하게 만들면 될 거 같슴다~!
수고하셨습니다!

Jungjjeong committed:
Fix: Comment Component child 형 추가

dong-onion commented on web/components/CommentInput/CommentInput.tsx:
```diff
@@ -0,0 +1,36 @@
+import Button from '../Button';
+import * as S from './CommentInput.style';
+
+const CommentInput = () => {
+  const ref = useRef<HTMLTextAreaElement>(null);
+  const [value, setValue] = useState('');
+
+  const handleChange = ({ target }: ChangeEvent<HTMLTextAreaElement>) => {
+    setValue(target.value);
+  };
+
+  const handleClick = () => {
+    console.log(value, '클릭');
+  };
+
```
textarea의 onChange로 state가 바뀔때 마다 리렌더링이 일어납니다.

state와 onChange를 사용하지 않고

```
const handleClick = () => {
   console.log( ref.current!.value , '클릭')
}
```

방식으로 comment value를 추출 하는게 어떨까 싶습니다

Jungjjeong replied:
> textarea의 onChange로 state가 바뀔때 마다 리렌더링이 일어납니다.
> 
> state와 onChange를 사용하지 않고
> 
> ```
> const handleClick = () => {
>    console.log( ref.current!.value , '클릭')
> }
> ```
> 
> 방식으로 comment value를 추출 하는게 어떨까 싶습니다

그러게요 불필요한 리렌더링이 많이 발생하는 군요..
실시간으로 바뀌는 글자 갯수를 렌더링하기 위해서 `onChange`를 사용했습니다. 
![image](https://user-images.githubusercontent.com/72294509/182301931-6200f7c4-6f72-4e2f-b251-280a062d8ed0.png)
ref를 사용하면 실시간으로 바뀌는 글자 갯수를 렌더링하지 못합니다. 
어떻게 해결할 수 있을까요? 😂

Miral3 replied:
다른곳이 렌더링 되는게 아니라 CommentInput 컴포넌트만 렌더링 되기도 하고, 글자수를 입력이 일어날 때마다 체크 하려면 리렌더링은 어쩔 수 없다고 생각합니다.

dong-onion reviewed (commented):
어려운 부분을 맡고도 많은양을 해내시느라 정말 고생많으셨습니다!
디자인도 너무 깔끔하고 잘 구현하신것 같습니다!

Miral3 commented on web/components/Comment/Comment.style.ts:
```diff
@@ -0,0 +1,62 @@
+  }
+`;
+
+export const RepliesContainer = styled.div`
```
대댓글 사이에도 구분선이 있으면 좋을것 같습니다.
<img width="685" alt="image" src="https://user-images.githubusercontent.com/59829606/182300593-23b11ad3-d6ba-44d5-9c2a-238e62c4004c.png">

Miral3 commented on web/components/Comment/Comment.style.ts:
```diff
@@ -0,0 +1,62 @@
+  gap: 10px;
+`;
+
+export const RepliesButton = styled.button`
```
각 버튼에 `cursor:pointer`를 줘서 클릭할 수 있다는걸 알려주면 좋을것 같습니다!

Miral3 commented on web/components/CommentInput/CommentInput.tsx:
```diff
@@ -0,0 +1,36 @@
+      <S.InputContainer>
+        <S.TextInput
+          placeholder="댓글을 입력해주세요."
+          ref={ref}
```
ref는 어떤 용도로 사용하는 건가요??

Jungjjeong replied:
textarea 값을 비우는 용도로 만들어놨습니다.

Miral3 reviewed (commented):
지영님 컴포넌트 구현하시느라 고생하셨습니다.
코드를 깔끔하게 잘 짜신것 같습니다. 스타일 관련되서 제 의견을 남겼습니다 확인 부탁드릴게요.
질문 남기신 구조분해 할당 같은 경우는 parent의 id와 child의 id가 혼동될 수 있기 때문에 child.id와 같은 방식으로 작성하는게 맞는것 같습니다.

Jungjjeong committed:
Refactor: 코드리뷰 반영 (Design + child 속성)

- Design : 구분선 추가, 버튼 cursor: pointer
- Refactor: child 속성 구조 분해 할당이 아닌, child.id와 같은 형식으로 작성
